bundles/specmate-connectors/src/com/specmate/connectors/internal/ConnectorUtil.java

```diff
@@ -0,0 +1,304 @@
+	 * BATCH_SIZE
+	 */
+	public void syncRequirementsFromSources() {
+		lock.lock();
+		try {
+			logService.log(LogService.LOG_INFO, "Synchronizing requirements");
+			Resource resource = transaction.getResource();
+			for (IRequirementsSource source : requirementsSources) {
+				logService.log(LogService.LOG_INFO, "Retrieving requirements from " + source.getId());
+				try {
+					Collection<Requirement> requirements = source.getRequirements();
+					if (requirements == null) {
+						continue;
+					}
+
+					IContainer localRootContainer = getOrCreateLocalRootContainer(resource, source.getId());
+					// Build hashset (extid -> requirement) for local requirements
+					TreeIterator<EObject> localIterator = localRootContainer.eAllContents();
+					HashMap<String, EObject> localRequirementsMap = new HashMap<>();
+					buildExtIdMap(localIterator, localRequirementsMap);
+
+					Requirement[] reqArray = requirements.toArray(new Requirement[0]);
+					int greatestUnhandledIndex = 0;
+					int maxIndex = requirements.size() - 1;
+					while (greatestUnhandledIndex <= maxIndex) {
+						int upperIndexExclusive = Math.min(greatestUnhandledIndex + BATCH_SIZE, maxIndex + 1);
+						Requirement[] current = Arrays.copyOfRange(reqArray, greatestUnhandledIndex,
+								upperIndexExclusive);
+						greatestUnhandledIndex = upperIndexExclusive;
+						List<Requirement> tosync = Arrays.asList(current);
+
+						try {
+							transaction.doAndCommit(new IChange<Object>() {
+								@Override
+								public Object doChange() throws SpecmateException {
+									syncContainers(localRootContainer, localRequirementsMap, tosync, source);
+									return null;
+								}
+							});
+						} catch (Exception e) {
+							logService.log(LogService.LOG_ERROR,
+									"An error occured while committing synced requirements. Reason:" + e.getMessage(),
+									e);
+							transaction.rollback();
+						}
+
+					}
+				} catch (Exception e) {
+					logService.log(LogService.LOG_ERROR,
+							"An error occured synching requirements. Reason: " + e.getMessage(), e);
+					transaction.rollback();
+				}
+			}
+		} finally {
+			lock.unlock();
+		}
+	}
+
+	/**
+	 * Retrieves a single requirement with a given id
+	 */
+	public void syncRequirementById(String id) {
+		lock.lock();
+		try {
+			logService.log(LogService.LOG_INFO, "Synchronizing requirement by id");
+			Resource resource = transaction.getResource();
+			for (IRequirementsSource source : requirementsSources) {
+				logService.log(LogService.LOG_INFO, "Retrieving single requirement from " + source.getId());
+				try {
+					Requirement requirement = source.getRequirementById(id);
+					if (requirement == null) {
+						continue;
+					}
+
+					IContainer localRootContainer = getOrCreateLocalRootContainer(resource, source.getId());
+					// Build hashset (extid -> requirement) for local requirements
+					TreeIterator<EObject> localIterator = localRootContainer.eAllContents();
```

Comment:
There seems to be a lot of redundancy to the syncRequirements method -> please refactor